File: examples/python/basic_usage.py
#!/usr/bin/env python3
"""
Example: Load the CMS sample database and demonstrate common queries.

Usage:
    python basic_usage.py
    python basic_usage.py --url https://raw.githubusercontent.com/<user>/the-barn/main/data/samples_2017UL.json
"""

import sys
import argparse

# ── sample_loader is in ../../python/ ────────────────────────
sys.path.insert(0, "../../python")
from sample_loader import SampleDB


def main():
    parser = argparse.ArgumentParser(description="The Barn — example usage")
    parser.add_argument("--json", default="../../data/samples_2017UL.json")
    parser.add_argument("--url",  default=None, help="Load from remote URL instead")
    args = parser.parse_args()

    # ── 1. Load ──────────────────────────────────────────────
    if args.url:
        db = SampleDB.from_url(args.url)
        print(f"Loaded from URL: {args.url}")
    else:
        db = SampleDB.from_json(args.json)
        print(f"Loaded from file: {args.json}")

    print(db.summary())
    print()

    # ── 2. Single sample access ──────────────────────────────
    tt = db["TTTo2L2Nu_TuneCP5_powheg"]
    print("[Single sample]")
    print(f"  Name    : {tt.name}")
    print(f"  DAS     : {tt.das_path}")
    print(f"  xsec    : {tt.cross_section_pb} pb")
    print(f"  N_evt   : {tt.nevents:,}")
    print(f"  N_eff   : {tt.effective_nevents:,.0f}")
    print(f"  accuracy: {tt.accuracy}")
    if tt.comment:
        print(f"  comment : {tt.comment}")
    print()

    # ── 3. Filter by substring ───────────────────────────────
    dy = db.filter("DYJetsToLL")
    print(f"[Filter: DYJetsToLL] — {len(dy)} matches")
    print(f"  {'Name':<50} {'xsec [pb]':>12} {'N events':>14}")
    print("-" * 78)
    for s in dy:
        xsec = f"{s.cross_section_pb:>12.1f}" if s.cross_section_pb else f"{'(pending)':>12}"
        nev  = f"{s.nevents:>14,}" if s.nevents else f"{'(pending)':>14}"
        print(f"  {s.name:<50} {xsec} {nev}")
    print()

    # ── 4. Filter by process prefix ──────────────────────────
    top = db.filter_by_process("TT", "ST_")
    print(f"[Top samples (TT* + ST_*)] — {len(top)} matches")
    for s in top:
        xsec = f"{s.cross_section_pb:.1f} pb" if s.cross_section_pb else "pending"
        print(f"  {s.name:<55} {xsec}")
    print()

    # ── 5. Typical analysis: luminosity weight ───────────────
    target_lumi = 41.5  # fb^-1 for 2017
    print(f"[Lumi weight example] target = {target_lumi} fb^-1")
    for name in ["WJetsToLNu_TuneCP5_amcatnloFXFX", "TTTo2L2Nu_TuneCP5_powheg"]:
        s = db[name]
        if s.cross_section_pb and s.nevents:
            neff = s.effective_nevents or s.nevents
            weight = (s.cross_section_pb * target_lumi * 1000) / neff
            print(f"  {name}")
            print(f"    w = ({s.cross_section_pb} × {target_lumi}k) / {neff:.0f} = {weight:.6f}")


if __name__ == "__main__":
    main()

File: examples/cpp/basic_usage.cpp
/**
 * @file basic_usage.cpp
 * @brief Demonstrates how to use the CMS SampleDB.
 *
 * Build:
 *   make                                   (uses the provided Makefile)
 *   ./bin/basic_usage ../../data/samples_2017UL.json
 */

#include "cms/sample_loader.h"

#include <iomanip>
#include <iostream>
#include <string>

static void print_divider(int w = 80) { std::cout << std::string(w, '-') << "\n"; }

static void print_row(const cms::SampleInfo& s) {
    std::cout << std::left  << std::setw(52) << s.name
              << std::right << std::setw(14);
    if (s.cross_section_pb) std::cout << *s.cross_section_pb;
    else                    std::cout << "(pending)";
    std::cout << std::setw(14);
    if (s.nevents) std::cout << *s.nevents;
    else           std::cout << "(pending)";
    std::cout << "\n";
}

int main(int argc, char* argv[]) {
    std::string json_path = (argc > 1)
        ? argv[1]
        : "../../data/samples_2017UL.json";

    try {
        // ── 1. Load ─────────────────────────────────────────
        auto db = cms::SampleDB::from_json(json_path);
        std::cout << "\n=== The Barn — CMS Sample Database ===\n"
                  << "Loaded : " << db.size() << " samples ("
                  << db.count_filled() << " with cross section)\n\n";

        // ── 2. Single sample access ─────────────────────────
        if (auto* tt = db.find("TTTo2L2Nu_TuneCP5_powheg"); tt) {
            std::cout << "[Single sample]\n"
                      << "  Name    : " << tt->name << "\n"
                      << "  DAS     : " << tt->das_path << "\n"
                      << "  xsec    : " << tt->cross_section_pb.value_or(-1) << " pb\n"
                      << "  N_evt   : " << tt->nevents.value_or(0) << "\n"
                      << "  N_eff   : " << tt->effective_nevents().value_or(0) << "\n"
                      << "  accuracy: " << tt->accuracy.value_or("N/A") << "\n\n";
        }

        // ── 3. Filter by pattern ────────────────────────────
        auto dy = db.filter("DYJetsToLL");
        std::cout << "[Filter: DYJetsToLL] — " << dy.size() << " matches\n";
        print_divider();
        std::cout << std::left  << std::setw(52) << "Name"
                  << std::right << std::setw(14) << "xsec [pb]"
                  << std::setw(14) << "N events" << "\n";
        print_divider();
        for (const auto* s : dy) print_row(*s);
        std::cout << "\n";

        // ── 4. Iterate all top samples ──────────────────────
        std::cout << "[Top-quark samples (TT* + ST_*)]\n";
        print_divider();
        for (const auto& [name, info] : db) {
            if (name.find("TT") == 0 || name.find("ST_") == 0)
                print_row(info);
        }
        std::cout << "\n";

        // ── 5. Safe access (missing sample) ─────────────────
        if (!db.contains("SomeHypotheticalSample")) {
            std::cout << "[Safe access] 'SomeHypotheticalSample' not found — skipped.\n";
        }

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    return 0;
}
